Switch Operator to use *metrics.Registry infra.

tommyp1ckles opened the pull request:
With the changes in https://github.com/cilium/cilium/pull/38743, the operator now has the Hive infra modules necessary to run the `shell` command.  This gives users access to a interactive shell that allows users to run included hive commands.

One thing that is missing is the `metrics` commands, which provide a buffered in-memory set of prometheus metrics.  This is because the operator still sets up its own metrics infra rather than using the `pkg/metrics.Registry` type.

This PR makes the necessary changes such that we can begin to use the Registry in the operator.


```release-note
add support for `metrics` in the cilium-operator when using the operator shell command. 
```

tommyp1ckles committed:
metrics: fix help message with valid command.

Running the command as is will result in:
<stdin>:0: metrics/plot -rate go.*heap_alloc_bytes: unknown shorthand flag: 'r' in -rate

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
pkg/metrics: prepare *metrics.Registry for use by operator.

1) Add ctor for *metrics.Registry type without agent metrics.
2) Expose underlying registry gather functionality needed by
	operator metrics command.
3) Add OperatorCell which provides minimal metrics infra for
	Operator to use:
	i) Allow operator to initialize lifecycle hooks manually
		to allow us to manually gather and register operator
		metrics.

*metrics.Registry is used by the agent hive to register metrics.
This does a lot of agent specific stuff when initializing.
In order to make this more generic, we create a agent specific
constructor and a general one that does not handle any initialization.

This allows us to create a OperatorCell that provides the registry
and allows the operator metrics code to initialize its metrics
manually.

We will use this to replace the operator metric init such
that we're basing it on the *Registry type.

The benefits of this are that this will allow us to include
the metrics commands types which will be very useful in conjunction
with recent commits to add the `shell` command to the operator binary.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
operator/metrics: use metrics.OperatorCell to init metrics.

This will allow us to use the `metrics` commands and sampler
types in the operator.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
ipam: use *metrics.Registry for IPAM metrics.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

gandro reviewed (approved):
My codeowners look good to me. Note that there are multiple compliation errors however.

doniacld reviewed (approved):
Approved for my codeowners group!

aanm commented on operator/metrics/node/cilium_node_metrics.go:
```diff
@@ -0,0 +1,139 @@
+	// If neither kvstore or node manager watchers are enabled (i.e. if the
+	// operator does not need to watch for them in the specified ipam mode
+	// then we just provide a noop metric set that we expect to never be used.
+	if !option.Config.IsNodeManagerEnabled() && !operatorOption.Config.IsKVstoreEnabled() {
```
Why do we have this file under the `operator/` This code can be reused for both the agent and the operator, no?

tommyp1ckles replied:
I think I'm going to drop this code altogether, the fact that this is setup separately from the other workqueue metrics seems to cause the cilium_ prefix to not work.  Instead I'll try to fix all of the watcher workqueue metrics in operator so that they start with `cilium_operator_`

tommyp1ckles committed:
k8s/watchers/metrics: register workqueue metrics directly.

As previous comments explain - the controller runtime SetProvider
is prone to racing, such that if the order changes and the init
inside k8s controller runtime libraries is runs first then the
SetProvider call inside k8s/watchers does nothing.

To improve this, we now export a global provider and attempt to
configure all k8s watcher work queues directly rather than
the implicit global provider.

To facilitate this, pkg/k8s/watchers/metrics.go is moved to a
new package to avoid import cycles.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tamilmani1989 reviewed (approved):
lgtm on behalf of sig-azure

tommyp1ckles committed:
api/metrics: fix tests

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
Documentation: update operator docs following using metrics registry

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

derailed reviewed (approved):
@tommyp1ckles Nice work!

tommyp1ckles committed:
clustermesh: ignore goleak for workqueue metrics collection.

There is precedent for ignoring this in other tests.
This test would otherwise pass if it is allowed to sleep
for a second following the test completing, indicating
that the goroutine won't actually be leaked.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

MrFreezeex reviewed (approved):
Thanks, clustermesh area looks good!
